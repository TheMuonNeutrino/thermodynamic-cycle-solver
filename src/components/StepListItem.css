.step-list-dropdown-wrapper{
    padding-left: 0.5em;
    padding-right: 0.5em;
}

.item-title{
    margin-bottom: 0.3em !important;
}