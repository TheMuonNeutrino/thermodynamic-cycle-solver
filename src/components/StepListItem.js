import React, { useEffect } from 'react';
import ParameterField from './ParameterField';
import { Button, Form } from 'semantic-ui-react';
import './StepListItem.css';
import { hasDefinedKey } from '../Utils';

const DropDown = ({value,setValue}) => {

    const dropdownOptions = [
        {key: 'none', text:'(None)', value: 'none'},
        {key: 'isobaric', text:'Isobaric', value:'isobaric'},
        {key: 'isochoric', text:'Isochoric', value:'isochoric'},
        {key: 'isothermal', text:'Isothermal', value:'isothermal'},
        {key: 'isentropic', text:'Isentropic', value:'isentropic'},
    ]

    return(
        <Form.Dropdown 
            search selection 
            options={dropdownOptions} 
            label='Step Type' 
            onChange={(e,data)=>{setValue(data.value)}}
            value={value}
            className='no-drag'
        />
    )
}

const StepListItem = ({step,index,setStep,deleteStep,deleteDisabled}) => {

    var entropyReadOnly = true

    if (hasDefinedKey(step,'staticEntropy')){
        step.entropy = step.staticEntropy
        entropyReadOnly = false
    }
    if (!hasDefinedKey(step,'entropy')){
        step.entropy = ''
    }

    if (!hasDefinedKey(step,'entropyChange')){
        step.entropyChange = ''
    }

    const stepUpdateFunction = (param) =>{
        setStep(index,param)
    }

    var deleteButtonParams = {
        size: 'small',
        icon: 'trash',
        onClick: ()=>{deleteStep(index)},
        className: 'no-drag',
    }

    if (deleteDisabled){
        deleteButtonParams.disabled = true
    }else{
        deleteButtonParams.negative = true
    }

    console.log(deleteDisabled,deleteButtonParams)
    

    const firstLinefieldProperties = [
        {
            label: 'Pressure',
            key: 'pressure',
            readOnly: false,
            updateKey: 'pressure',
        },{
            label: 'Volume',
            key: 'volume',
            readOnly: false,
            updateKey: 'volume',
        },{
            label: 'Temperature',
            key: 'temperature',
            readOnly: false,
            updateKey: 'temperature',
        },{
            label: 'Entropy',
            key: 'entropy',
            readOnly: entropyReadOnly,
            updateKey: 'staticEntropy'
        }
    ]

    function numToSSColumn(num){
        let s = '', t;
      
        while (num > 0) {
          t = (num - 1) % 26;
          s = String.fromCharCode(65 + t) + s;
          num = (num - t)/26 | 0;
        }
        return s || undefined;
      }

    return(
        <div className='ui container'>
            <div className='ui large form'>
                <div>
                    <h4 className='item-title'>
                        <i className='pad-right'>Point {numToSSColumn(index+1)}</i>
                        <Button 
                            {...deleteButtonParams}
                        ></Button>
                    </h4>
                </div>
                <div className='fields'>
                    {firstLinefieldProperties.map((item)=>{return(
                        <ParameterField 
                            label={item.label}
                            value={step[item.key]}
                            key={item.label}
                            readOnly={item.readOnly}
                            updateValue={stepUpdateFunction}
                            updateKey={item.updateKey}
                        />
                    )})}
                </div>
                <div className='fields'>
                    <div className='step-list-dropdown-wrapper'>
                        <DropDown 
                            value={step.type}
                            setValue={(value)=>{stepUpdateFunction({type: value})}}
                        />
                    </div>
                    <ParameterField
                        label='Entropy Change'
                        value={step.entropyChange}
                        key='entropyChange'
                        readOnly={true}
                        updateKey='entropyChange'
                        updateValue={stepUpdateFunction}
                    />
                </div>
                <hr/>
            </div>
        </div>
    )
}

export default StepListItem