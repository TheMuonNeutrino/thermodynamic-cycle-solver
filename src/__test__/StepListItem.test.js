import React from 'react';
import {mount} from 'enzyme';
import StepListItem from '../components/StepListItem';
import {createSerializer} from 'enzyme-to-json';
expect.addSnapshotSerializer(createSerializer({mode: 'deep'}));

describe.skip('StepList tests',()=>{
    var step, index, setStep, wrapper
    beforeEach(()=>{
        step = {
            temperature: 100,
            pressure: 1000000,
            volume: 5,
            staticEntropy: 5,
            type: 'isobaric',
            entropyChange: 300,
        }
        index = 3
        setStep = jest.fn()
        wrapper = mount(
            <StepListItem step={step} index={index} setStep={setStep}/>
        )
    })
    it('should render StepList',()=>{
        expect(wrapper).toMatchSnapshot();
    })
    it('should pass changes to the first 4 fields to setStep, parsing floats',()=>{
        wrapper.find('.parameter-field-value').at(0).simulate('change',{
            target: {value: '10'} 
        })
        expect(setStep).toHaveBeenLastCalledWith(index,{pressure: 10})
        wrapper.find('.parameter-field-value').at(2).simulate('change',{
            target: {value: '1e5'} 
        })
        expect(setStep).toHaveBeenLastCalledWith(index,{temperature: 100000})
    })
    it('should reject none float inputs',()=>{
        wrapper.find('.parameter-field-value').at(0).simulate('change',{
            target: {value: 'abc'} 
        })
        expect(setStep).not.toHaveBeenCalled()
    })
    it('should pass changes to step type to setStep',()=>{
        wrapper.find('span.text').at(2).simulate('click')
        expect(setStep).toHaveBeenLastCalledWith(3,{type: 'isochoric'})
    })
})