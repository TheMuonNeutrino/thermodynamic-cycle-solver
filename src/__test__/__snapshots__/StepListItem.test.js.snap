// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StepList tests should render StepList 1`] = `
<div
  className="ui container"
>
  <div
    className="ui large form"
  >
    <div
      className="fields"
    >
      <div
        className="field parameter-field"
      >
        <label
          className="parameter-field-label"
        >
          Pressure
        </label>
        <input
          className="parameter-field-value"
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          readOnly={false}
          type="text"
          value={1000000}
        />
      </div>
      <div
        className="field parameter-field"
      >
        <label
          className="parameter-field-label"
        >
          Volume
        </label>
        <input
          className="parameter-field-value"
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          readOnly={false}
          type="text"
          value={5}
        />
      </div>
      <div
        className="field parameter-field"
      >
        <label
          className="parameter-field-label"
        >
          Temperature
        </label>
        <input
          className="parameter-field-value"
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          readOnly={false}
          type="text"
          value={100}
        />
      </div>
      <div
        className="field parameter-field"
      >
        <label
          className="parameter-field-label"
        >
          Entropy
        </label>
        <input
          className="parameter-field-value"
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          readOnly={false}
          type="text"
          value={5}
        />
      </div>
    </div>
    <div
      className="fields"
    >
      <div
        className="step-list-dropdown-wrapper"
      >
        <div
          className="field"
        >
          <label>
            Step Type
          </label>
          <div
            aria-describedby={null}
            aria-expanded={false}
            className="ui search selection dropdown"
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onMouseDown={[Function]}
            role="combobox"
          >
            <input
              aria-autocomplete="list"
              autoComplete="off"
              className="search"
              onChange={[Function]}
              style={
                Object {
                  "width": undefined,
                }
              }
              tabIndex={0}
              type="text"
              value=""
            />
            <div
              aria-atomic={true}
              aria-live="polite"
              className="divider text"
              role="alert"
            >
              Isobaric
            </div>
            <i
              aria-hidden="true"
              className="dropdown icon"
              onClick={[Function]}
            />
            <div
              className="menu transition"
              role="listbox"
            >
              <div
                aria-checked={false}
                aria-selected={false}
                className="item"
                onClick={[Function]}
                role="option"
                style={
                  Object {
                    "pointerEvents": "all",
                  }
                }
              >
                <span
                  className="text"
                >
                  (None)
                </span>
              </div>
              <div
                aria-checked={true}
                aria-selected={true}
                className="active selected item"
                onClick={[Function]}
                role="option"
                style={
                  Object {
                    "pointerEvents": "all",
                  }
                }
              >
                <span
                  className="text"
                >
                  Isobaric
                </span>
              </div>
              <div
                aria-checked={false}
                aria-selected={false}
                className="item"
                onClick={[Function]}
                role="option"
                style={
                  Object {
                    "pointerEvents": "all",
                  }
                }
              >
                <span
                  className="text"
                >
                  Isochoric
                </span>
              </div>
              <div
                aria-checked={false}
                aria-selected={false}
                className="item"
                onClick={[Function]}
                role="option"
                style={
                  Object {
                    "pointerEvents": "all",
                  }
                }
              >
                <span
                  className="text"
                >
                  Isothermal
                </span>
              </div>
              <div
                aria-checked={false}
                aria-selected={false}
                className="item"
                onClick={[Function]}
                role="option"
                style={
                  Object {
                    "pointerEvents": "all",
                  }
                }
              >
                <span
                  className="text"
                >
                  Isentropic
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="field parameter-field"
      >
        <label
          className="parameter-field-label"
        >
          Entropy Change
        </label>
        <input
          className="parameter-field-value"
          disabled={true}
          onBlur={[Function]}
          onChange={[Function]}
          readOnly={true}
          type="text"
          value={300}
        />
      </div>
    </div>
  </div>
</div>
`;
